explorer/interpreter/type_checker.cpp

```diff
@@ -4232,6 +4222,49 @@ auto TypeChecker::DeclareInterfaceDeclaration(
   for (Nonnull<Declaration*> m : iface_decl->members()) {
     CARBON_RETURN_IF_ERROR(DeclareDeclaration(m, iface_scope_info));
 
+    // For an `extends C;` declaration, add `Self is C` to our constraint.
+    if (auto* extends = dyn_cast<InterfaceExtendsDeclaration>(m)) {
+      CARBON_ASSIGN_OR_RETURN(Nonnull<const Value*> base,
+                              TypeCheckTypeExp(extends->base(), iface_scope));
+      auto* constraint_type = dyn_cast<ConstraintType>(base);
+      if (auto* interface_type = dyn_cast<InterfaceType>(base)) {
+        CARBON_ASSIGN_OR_RETURN(
+            constraint_type,
+            MakeConstraintForInterface(extends->source_loc(), interface_type));
+      }
+      if (!constraint_type) {
+        return ProgramError(extends->source_loc())
+               << "an interface can only extend a constraint, found " << *base;
+      }
+      CARBON_RETURN_IF_ERROR(builder.AddAndSubstitute(
+          *this, constraint_type, builder.GetSelfType(),
+          builder.GetSelfWitness(), Bindings(),
+          /*add_lookup_contexts=*/true));
+    }
+
+    // For an `impl X as Y;` declaration, add `X is Y` to our constraint.
+    if (auto* impl = dyn_cast<InterfaceImplDeclaration>(m)) {
```

Comment:
Can this and line 4268 be an `else if`?

Maybe this should be rewritten as a switch on m->Kind(), since there are now three different types being checked for?

Reply:
Sure, let's make this a `switch`. As discussed offline, I think we probably want to merge this back into `DeclareDeclaration` somehow in the longer term.